lib/AST/ASTDemangler.cpp

```diff
@@ -662,6 +663,45 @@ Type ASTBuilder::createSILBoxType(Type base) {
   return SILBoxType::get(base->getCanonicalType());
 }
 
+Type ASTBuilder::createSILBoxTypeWithLayout(
+    const std::vector<BuiltSILBoxField> &fields,
+    const std::vector<BuiltSubstitution> &Substitutions,
+    const std::vector<BuiltRequirement> &Requirements) {
+  SmallVector<Type, 4> replacements;
+  SmallVector<GenericTypeParamType *, 4> genericTypeParams;
+  for (const auto &s : Substitutions) {
+    if (auto *t = dyn_cast_or_null<GenericTypeParamType>(s.first.getPointer()))
+      genericTypeParams.push_back(t);
+    replacements.push_back(s.second);
+  }
+
+  GenericSignature signature;
+  if (!genericTypeParams.empty())
+    signature = GenericSignature::get(genericTypeParams, Requirements);
+  SmallVector<SILField, 4> silFields;
+  for (auto field: fields)
+    silFields.emplace_back(field.getPointer()->getCanonicalType(),
+                           field.getInt());
+  SILLayout *layout =
+      SILLayout::get(Ctx, signature.getCanonicalSignature(), silFields);
+  SmallVector<ProtocolConformanceRef, 4> conformanceRefs;
+  llvm::SmallDenseMap<Type, ProtocolConformanceRef, 4> conformance;
+  for (const auto &r : Requirements)
+    if (r.getKind() == RequirementKind::Conformance)
+      if (auto *proto =
+              llvm::dyn_cast<ProtocolDecl>(r.getSecondType()->getAnyNominal()))
+        conformance.insert({r.getFirstType(), ProtocolConformanceRef(proto)});
+
+  for (const auto &param : genericTypeParams)
+    if (const auto proto = conformance.lookup(param))
+      conformanceRefs.push_back(proto);
```

Comment:
The elements of the conformanceRefs array are in one-to-one correspondence with protocol conformance requirements of the generic signature, not the generic parameters. Also you'll need to do the actual conformance lookup here instead of constructing a ProtocolConformanceRef.

Instead, you should call SubstitutionMap::get() with a pair of callbacks:

- One callback returns the right replacement type for the generic parameter
- For the other callback, pass in an instance of the LookUpConformanceInModule() functor.

Reply:
Done!